core/src/odbc_uri.rs

```diff
@@ -754,11 +791,32 @@ mod unit {
             use crate::odbc_uri::ODBCUri;
             let expected_opts =
                 ClientOptions::parse("mongodb://foo2:bar2@127.0.0.2:27017").unwrap();
-            let opts = ODBCUri::new("SERVER=127.0.0.2:27017;URI=mongodb://foo:bar@127.0.0.1:27017")
+            let opts = ODBCUri::new(
+                "SERVER=127.0.0.2:27017;URI=mongodb://foo:bar@127.0.0.1:27017".to_string(),
+            )
+            .unwrap()
+            .try_into_client_options()
+            .unwrap();
+            assert_eq!(expected_opts.hosts[0], opts.hosts[0]);
+        }
+
+        #[test]
+        fn supplied_args_override_dsn_args() {
+            // we leave dsn out of the uri so that it doesn't try to query the registry in the test
+            use crate::odbc_uri::ODBCUri;
+            let expected_opts = ClientOptions::parse(
+                "mongodb://foo:bar@www.atlas.net:27017/?authSource=authDB&ssl=true",
+            )
+            .unwrap();
+            let expected_cred = expected_opts.credential.unwrap();
+            let opts = ODBCUri::new("UID=foo;PWD=bar;SERVER=www.atlas.net:27017;User=foo2;Password=bar2;uri=mongodb://localhost:29000/?authSource=authDB&ssl=true".to_string())
```

Comment:
Would you provide a little more context on this test?  
Is it to confirm that the options specified earlier in the string are not overridden by the ones specified later?

Reply:
Yep, that's correct! If a DSN is encountered in the connection string, it will be fetched and appended to the end of the connection string, and re-processed. However, we want earlier options to be honored first so that users can override existing dsn settings.

Reply:
Ah cool, and the expected_opts is simply a way to get the `user:foo`  `pass:bar` credentials to check against.  Got it thanks!